delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/auth/AuthTokenProvider.scala

```diff
@@ -0,0 +1,74 @@
+package ch.epfl.bluebrain.nexus.delta.sdk.auth
+
+import cats.effect.Clock
+import ch.epfl.bluebrain.nexus.delta.kernel.cache.KeyValueStore
+import ch.epfl.bluebrain.nexus.delta.kernel.utils.IOUtils
+import ch.epfl.bluebrain.nexus.delta.sdk.auth.AuthMethod.{Anonymous, Credentials}
+import ch.epfl.bluebrain.nexus.delta.sdk.identities.model.AuthToken
+import monix.bio.UIO
+
+import java.time.{Duration, Instant}
+
+/**
+  * Provides an auth token for the service account, for use when comunicating with remote storage
+  */
+trait AuthTokenProvider {
+  def apply(): UIO[Option[AuthToken]]
+}
+
+object AuthTokenProvider {
+  def apply(auth: Option[AuthMethod], keycloakAuthService: KeycloakAuthService): AuthTokenProvider = {
+    auth match {
+      case Some(credentials: Credentials)    => new CachingKeycloakAuthTokenProvider(credentials, keycloakAuthService)
+      case Some(AuthMethod.AuthToken(token)) => new FixedAuthTokenProvider(AuthToken(token))
+      case Some(Anonymous) | None            => new AnonymousAuthTokenProvider
+    }
+  }
+  def anonymousForTest: AuthTokenProvider = new AnonymousAuthTokenProvider
+}
+
+private class AnonymousAuthTokenProvider extends AuthTokenProvider {
```

Comment:
Passing the `credentials` as a parameter also remove the need of subclassing here

Reply:
I don't understand, anonymous doesn't need credentials?

Reply:
See previous comment